File: Data Structures & Algorithms/sliding-window-maximum/submission-1.rs
use std::collections::VecDeque;

impl Solution {
    pub fn max_sliding_window(nums: Vec<i32>, k: i32) -> Vec<i32> {
        // Pre-allocate to avoid reallocations
        let mut res = Vec::with_capacity(nums.len().saturating_sub(k as usize - 1));
        let mut deque: VecDeque<usize> = VecDeque::new();

        for (i, &num) in nums.iter().enumerate() {
            // 1. Remove the index at the front if it's no longer in the window
            if let Some(&first_idx) = deque.front() {
                if first_idx + (k as usize) <= i {
                    deque.pop_front();
                }
            }

            // 2. Pop smaller elements from the back (they are useless now)
            while let Some(&last_idx) = deque.back() {
                if nums[last_idx] <= num {
                    deque.pop_back();
                } else {
                    break;
                }
            }

            // 3. Add the current index
            deque.push_back(i);

            // 4. If our window has reached size k, record the max
            if i >= (k as usize) - 1 {
                // The front of the deque is always the maximum
                res.push(nums[*deque.front().unwrap()]);
            }
        }

        res
    }
}

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return 0
        current_max = 1
        current_min = 1
        res = nums[0]
        
        for n in nums:
            if n < 0:
                current_max, current_min = current_min, current_max

            current_max = max(n, n * current_max)
            current_min = min(n, n * current_min)

            res = max(res, current_max)

        return res

        
